### ISSUE TITLE
Containers: which go in Kokkos or Kokkos Kernels

### ISSUE THREAD

**Issue by lucbv** [kokkos/kokkos#4255]
@crtrott @srajama1 
During the Aug-18 stand-up we discussed the possibility of moving some containers from one repo to another. Below are the two containers previously discussed, if others should be added to the list, feel free to edit this issue.

- StaticCrsGraph: should it move in Kokkos Kernels? It is not used internally, the Kokkos::Graph is preferred for tasking.
- UniformMemoryPool: should it move to Kokkos? It was implemented to support the spgemm algorithm but users of Kokkos sometime build and link against Kokkos Kernels just for that feature...

**Comment by ajpowelsnl** [comment]
@lucbv - I've added to "Items to Discuss" for this week's Kokkos developer meeting.

**Comment by lucbv** [comment]
Thanks Amy : )

**Comment by ajpowelsnl** [comment]
Of course, @lucbv.  BTW, is everyone aware of all of the containers under consideration?

**Comment by lucbv** [comment]
We talked about it last week so I believe they are but we also do not have to vote on it this week if we want to expand the discussion. It's just good to have it out there for folks to debate.

**Comment by crtrott** [comment]
ExaMiniMD for example uses StaticCrsGraph to demonstrate CRS based neighborlists.

**Comment by fnrizzi** [comment]
Just my personal thought, and obviously there is a tradeoff to evaluate since it would add one more repo to manage, but what about creating a `kokkos-containers` repo? In terms of dependency stack, since the CrsGraph uses Views, one would have: 
```
 kernels 
---------
containers
---------
  core 
```
where the `containers` also contain the CrsMatrix. Seems pretty clean to me. As a user, it would be intuitive. 
Also, if I wanted to use a CrsMatrix somewhere, this would allow me to just having to handle `containers/core`. 
Re the pool, I personally would add to `core`. 
Just my personal opinion. I do know that handling another repo might be a pain (`containers` could be made a submodule of kernels for example to simplify some things).

**Comment by dalg24** [comment]
Francesco's suggestion https://github.com/kokkos/kokkos/issues/4255#issuecomment-905522733 is tempting, especially if `DualView` gets lost in the process

**Comment by stanmoore1** [comment]
>especially if DualView gets lost in the process

Are you re-writing LAMMPS? :)

**Comment by srajama1** [comment]
Folks, the CRS is CRSGraph is compressed row sparse matrix, which is a foundation for linear algebra. If that doesn't belong in Kokkos Kernel then I don't know what does. 

I am not sure why folks are so averse to adding a dependency on Kokkos Kernels. What am I missing? If it is a brand new product from some external source there is an inherent risk associated with adding a dependency to a new library. It is an SNL developed library, that is part of the Kokkos ecosystem, and follows the same process for everything. I believe ExaminiMD (and/or other mini-apps) could use the entire Kokkos ecosystem, Core for programming model, Kokkos Kernels for linear algebra data structures or other kernels, and tools for profiling, debugging etc.

We are also modularizing Kokkos Kernels next FY, so Examinimd will only compile a small portion of Kokkos Kernels (container like portion).

On the opposite I have a reservations on moving MemoryPool to Core. What is the point of this? Why can't folks who want to use MemoryPool should just depend on Kokkos Kernels?

The Containers repo proposed above is attractive, but lot of work.  We are moving to that model any way without having additional repositories (one repo, but allow partial compilation, tests etc.) anyway next year. Folks can just get containers submodule of Kokkos Kernels which could have CrGraph and MemoryPool.

**Comment by fnrizzi** [comment]
disclaimer: everything I saw below is meant to be *constructive* and not a critic in any way. Just another perspective.

> the CRS in CRSGraph is compressed row sparse matrix, which is a foundation for linear algebra. If that doesn't belong in Kokkos Kernel then I don't know what does.

I see your point. However, if I were (and I am) a user---and being picky---I would consider the CrsMatrix a container, which i personally would see as distinct from kernels. Obviously, we are kind of splitting hair, and talking about "where" something belongs to. But given that the names of the main components have to remain, `kernels` makes me think of linear algebra kernels which depend on stuff. The fact that the containers are *part* of kernels or a separate entity, I see that as a detail.

> We are also modularizing Kokkos Kernels next FY, and The Containers repo proposed above is attractive, but lot of work.

(1) I don't understand why it would be too hard to split things into: `core/containers/kernels/tutorials/tools/pykokkos`. 
I think using some sort of "submoduling" (does not have to be git submodule, can be some of the other similar git things)
would make this totally feasible. In the end, I don't see an extra `container` repo as causing so much overhead to maintain since it would be pretty much mirroring the build system of core (in my view). In some sense, doing such separation would at least cut to zero all debates of where something belongs to :) there is no room for interpretation anymore. 

(2) Taking this to an extreme, but the Kokkos ecosystem does not necessarily have to be fully split in separate repo. 
One other way to simplify all this, would be to make a single repo put everything inside (dependency is top-to-bottom):
```
kokkos:
  kernels
  pykokkos (putting this here since I assume it does not depend on kernels) 
  containers
  core
  tools (I think it goes here right?)
```
or even finer-grained:
```
kokkos:
  kernels-batched
  kernels-dense
  kernels-sparse
  pykokkos (putting this here since I assume it does not depend on kernels) 
  containers
  core
  tools (I think it goes here right?)
```
BUT such that users can *pick, build and use individual* pieces. 
```
cmake -DKOKKOS_ENABLE_CORE=ON ... 
cmake -DKOKKOS_ENABLE_KERNELS=ON ...  # this automatically turns core on or you pass a specific core to link 
cmake -DKOKKOS_ENABLE_KERNELS_SPARSE=ON ... 
cmake -DKOKKOS_ENABLE_CONTAINERS=ON ...  
```
no repo-level split, no room for debating, everything in one place. Personally, I like (1) better but wanted to bring this up too.

**Comment by crtrott** [comment]
How I see it kokkos/kokkos was meant for general purpose stuff useful to a wide array of different application domains, while KokkosKernels was math focused. That is why we put StaticCrsGraph into Kokkos/kokkos back in the day, because we explicitly discussed it with the respective teams (and back in the day Mark and I were the ones responsible for "KokkosKernels"/"Tpetra" which owned the on-node math stuff) and we considered StaticCrsGraph more general purpose than the name indicated. Maybe we should have renamed it to LOSSL (List-Of-Static-Sized-Lists) to make that clearer, but it really is just the equivalent to a pointers of pointers array, in a Count-Allocate-Fill-Use world enforced by massive parallelism, and no viable fast parallel allocation mechanism on the device. Considering that this data structure is really trivially simple (I mean after all it is just a struct with and offset array and a values array) I doubt that anyone would think: Ah let me get a new dependency instead of writing this on my own. In fact: we have two of those in kokkos/kokkos, since we needed the same data structure internally but didn't want to make the core subpackage depend on the containers subpackage.

**Comment by fnrizzi** [comment]
Even if that thing is trivially simple, if you need that in `core`, why then not make a "subcomponent" (whose name TBD) that contains all the lowest-level things that even core depends on? so there would be no need to copy this around right? 
For example: 
```
etc
containers (that use views)
core
my-lowest-level-thing (your trivially simple struct that depends only on C++) 
```
anyway, i don't mean to say copying it around it not ok, just that all this process (as well what belongs to what) seems to me a result of the naming/what-belongs-where.

**Comment by stanmoore1** [comment]
>Folks, the CRS in CRSGraph is compressed row sparse matrix, which is a foundation for linear algebra. If that doesn't belong in Kokkos Kernel then I don't know what does.

CRS is generally useful and not just limited to linear algebra, hence the argument to stay in Kokkos/containers. ExaMiniMD and SPARTA have zero linear algebra kernels but still use CRS.

**Comment by fnrizzi** [comment]
> we needed the same data structure internally but didn't want to make the core subpackage depend on the containers subpackage.

Could that imply that maybe `core` would benefit from being broken up a bit?

**Comment by crtrott** [comment]
No it doesn't indicate core needs breaking up. Its more the opposite, It indicates sometimes its hard to separate containers from the core programming model since one needs data structures to implement the parallelism stuff too. On the other hand containers need parallelism to implement things like deep copy, initialization and what not.

**Comment by mhoemmen** [comment]
> That is why we put StaticCrsGraph into Kokkos/kokkos back in the day, because we explicitly discussed it with the respective teams (and back in the day Mark and I were the ones responsible for "KokkosKernels"/"Tpetra" which owned the on-node math stuff) and we considered StaticCrsGraph more general purpose than the name indicated.

Right, it's a struct with a few typedefs in it that Christian wrote 2013-4-ish.  It's nothing anyone would write a paper or a proposal about.

These days, I'm in Stan's position of wanting to minimize software dependencies and reduce build time.

**Comment by PhilMiller** [comment]
As a potential split-the-baby proposal, how about exposing the `ListOfLists` name and implementation in Core, and then composing `StaticCrsGraph` with whatever linear-algebra-specific functionality it may offer in its interface on top of that?

More seriously, it seems like if Kernels is developed by an overlapping group of people, with the same policies and infrastructure, it would make sense to consolidate it into the same repository as Core and Containers, with whatever additional configure-/build-time modularization effort to leave out the parts any given dependent application isn't using.
